### PR TITLE
Added recursive sleep function

### PR DESCRIPTION
**Pull request by erik-leven**
Before we always added sleep time after each request to avoid a 429 response. Now we only sleep if we get a 429, and the time the MS sleep is incremented with 'sleep_incremental' for every successive time the response is 429.

In addition we added a test to sort out deactivated entities, to avoid decoding their images when they are not needed,

### CODE REVIEW

File: service/cvpartner.py [hunk 1]
```diff
@@ -219,6 +216,44 @@ def __put_custom_tag(self, url, entity):
             abort(req.status_code, req.json())
         return str(req.status_code)
 
+    def check_429(self, req, cv_url, headers):
```
Comment (meyhane) [inline service/cvpartner.py]: this is not checking any status rather than only 429. name is misleading to me.

File: service/cvpartner.py [hunk 2]
```diff
@@ -219,6 +216,44 @@ def __put_custom_tag(self, url, entity):
...
+        if req.status_code == 429:
+            if os.environ.get('sleep_increment') is not None:
```
Comment (meyhane) [inline service/cvpartner.py]:
```suggestion
            if not os.environ.get('sleep_increment'):
```

File: service/cvpartner.py [hunk 3]
```diff
@@ -219,6 +216,44 @@ def __put_custom_tag(self, url, entity):
...
+                return self.recursive_request(cv_url, headers, float(os.environ.get('sleep_increment')))                        
+            else:
+                logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+                raise AssertionError("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+        else:
+            logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+            raise AssertionError("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))        
+
+    def recursive_request(self, cv_url, headers, sleep_increment):
+        if os.environ.get('sleep_total') == None:
```
Comment (meyhane) [inline service/cvpartner.py]:
```suggestion
        if not os.environ.get('sleep_total'):
```

File: service/cvpartner.py [hunk 4]
```diff
@@ -219,6 +216,44 @@ def __put_custom_tag(self, url, entity):
...
+            if os.environ.get('sleep_increment') is not None:
+                return self.recursive_request(cv_url, headers, float(os.environ.get('sleep_increment')))                        
+            else:
+                logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+                raise AssertionError("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+        else:
+            logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+            raise AssertionError("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))        
+
+    def recursive_request(self, cv_url, headers, sleep_increment):
```
Comment (meyhane) [inline service/cvpartner.py]: i suggest calling this function _recurse_for_429_ as it describes better what id does

File: service/cvpartner.py [hunk 5]
```diff
@@ -109,26 +109,23 @@ def __get_all_cvs(self, path):
...
             if req.status_code != 200:
-                logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
-                raise AssertionError("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+                req = self.check_429(req, cv_url, headers)
+
             clean = json.loads(req.text)
             offset += len(clean)
             cv_url = os.environ.get("base_url") + "v3/cvs/"
             for entity in clean:
+                if entity.get('deactivated', ""):
```
Comment (meyhane) [inline service/cvpartner.py]:
```suggestion
                if entity.get('deactivated'):
```

**Review by meyhane — COMMENTED** [review]
1. I suppose 429 response would have some headers like _remaining-time-for-limit_ or _try-after_ . Has it been considered if one can figure out sleep time from them?
2. is this handling of 429 response not applicable to other get calls? Then, you could bundle                           
following block into a method and use it for all get requests?

```
            req = requests.get(url, headers=headers)

            if req.status_code != 200:
                req = self.check_429(req, cv_url, headers)
```

File: README.md [hunk 6]
```diff
@@ -18,6 +18,8 @@ This microservice needs information about path to the url of the next page (more
 "reference_post": Data to send with POST to reference_url
 "reference_url": Path for GET of references.
 "references_path": Path to references in return object from reference_url
+"sleep_increment": Incremental time (s) the MS sleeps for each 429 response. The total sleep time adds up until it reached 'sleep_max'.
```
Comment (meyhane) [inline README.md]: These are obselete now that you use response header, right?

File: service/cvpartner.py [hunk 7]
```diff
@@ -214,11 +200,51 @@ def __put_custom_tag(self, url, entity):
...
         return str(req.status_code)
 
+    def check_error(self, req, url, headers, method, data_type = None, data = None):
+        if req.status_code == 429:
+            return self.recursive_request(url, headers, req.headers.get('Retry-After'), method, data_type, data)                        
+        else:
+            logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+            raise AssertionError("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))        
+
+    def recursive_request(self, url, headers, retry_after, method, data_type = None, data = None):
```
Comment (meyhane) [inline service/cvpartner.py]: I suggest changing "data_type" to "json" to get rid of if-else on data_type.
Just suggest, though.

**Review by meyhane — APPROVED** [review]
Approving with suggestions.
Should be  tested with put, post and get etc.

**Comment by erik-leven** [comment]
Fixed comments from Egemen

File: README.md [hunk 8]
```diff
@@ -216,7 +215,8 @@ PUT user
       "entities_path": "values",
       "headers": "{'Accept':'application/json', 'Content-Type':'application/json', 'Authorization':'$SECRET(token)'}",
       "next_page": "next.href",
-      "sleep": "0.400",
+      "sleep_increment": 0.5,
+      "sleep_max": 20,
```
Comment (meyhane) [inline README.md]: not relevant anymore.

File: service/cvpartner.py [hunk 9]
```diff
@@ -214,11 +200,39 @@ def __put_custom_tag(self, url, entity):
...
+        if req.status_code == 429:
+            return self.recursive_request(url, headers, req.headers.get('Retry-After'), method, json, data)                        
+        else:
+            logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+            raise AssertionError("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))        
+
+    def recursive_request(self, url, headers, retry_after, method, json = None, data = None):
+        logger.info("Sleeping for %.2f seconds" % float(retry_after))
+        sleep(float(retry_after))
+        if method == 'get':
```
Comment (meyhane) [inline service/cvpartner.py]: suggestion for thie whole if-else block
```suggestion
        if method in('get', 'post', 'put'):
            req = requests.request(method=method, url=url, headers=headers, json=json, data=data)
```

File: README.md (hunk 8 above)
Comment (erik-leven) [reply README.md]: removed

File: service/cvpartner.py (hunk 9 above)
Comment (erik-leven) [reply service/cvpartner.py]: nice, updated

**Review by meyhane — APPROVED** [review]

**Pull request merged**